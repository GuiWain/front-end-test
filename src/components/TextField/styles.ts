import styled, { css, DefaultTheme } from 'styled-components'

import { TextFieldProps } from '.'

import { Eye, EyeSlash } from '@styled-icons/bootstrap/'
import { CheckCircle } from '@styled-icons/boxicons-solid/'

type IconPositionProps = Pick<TextFieldProps, 'iconPosition'>

type WrapperProps = Pick<TextFieldProps, 'disabled'> & { error?: boolean } & {
  alert?: 'alertSuccess' | 'alertWarning'
}

export const EyeIcon = styled(Eye).attrs({
  'data-testid': 'eye-icon'
})`
  cursor: pointer;
`
export const EyeSlashIcon = styled(EyeSlash).attrs({
  'data-testid': 'eye-slash-icon'
})`
  cursor: pointer;
`
export const CheckCircleIcon = styled(CheckCircle).attrs({
  'data-testid': 'check-circle-icon'
})`
  ${({ theme }) => css`
    color: ${theme.colors.successMedium};
  `}
`

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled, alert }) => css`
    ${!!alert && alert === 'alertSuccess'
      ? wrapperModifiers.alertSuccess(theme)
      : null}
    ${!!alert && alert === 'alertWarning'
      ? wrapperModifiers.alertWarning(theme)
      : null}
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    align-items: center;

    background: ${theme.colors.white} 0% 0% no-repeat padding-box;
    border: 0.1rem solid ${theme.colors.secondaryLight};
    border-radius: ${theme.border.smallRadius};
    padding: 0 0.8rem;

    height: 3.6rem;

    &:focus-within {
      box-shadow: 0rem 0.4rem 0.6rem #00000014;
      border: 0.1rem solid ${theme.colors.infoDark};
    }
  `}
`

export const Input = styled.input<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    color: ${theme.colors.secondaryDark};
    font: normal normal ${theme.font.regular} ${
    theme.font.sizes.xxsmall
  } / 2.8rem
      ${theme.font.family};

    padding: 0.8rem 0;
    padding-${iconPosition}: 0.4rem;
    background: transparent;
    border: 0;
    outline: none;
    width: ${iconPosition === 'right' ? `calc(100% - 2.2rem)` : `100%`};

    &:-webkit-autofill {
      -webkit-box-shadow: 0 0 0 2.4
      #EAEAEA inset;
      filter: none
    }
  `}
`

export const Label = styled.label`
  ${({ theme }) => css`
    font: normal normal ${theme.font.semiBold} ${theme.font.sizes.xsmall} / 4rem
      ${theme.font.family};

    letter-spacing: -0.006rem;
    text-align: left;
    color: ${theme.colors.secondaryDark};
    cursor: pointer;
  `}
`

export const Icon = styled.div<IconPositionProps>`
  ${({ theme, iconPosition }) => css`
    display: flex;
    color: ${theme.colors.secondaryMedium};
    order: ${iconPosition === 'right' ? 1 : 0};

    & > svg {
      width: 2rem;
      height: 100%;
    }
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.errorMedium};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`

export const alertSuccess = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.successMedium};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`

export const AlertWarning = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.warningDark};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.errorMedium};

      &:focus-within {
        border: 0.1rem solid ${theme.colors.errorMedium};
      }
    }

    ${Icon},
    ${Label} {
      color: ${theme.colors.errorMedium};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label},
    ${Input},
    ${Icon} {
      cursor: not-allowed;
      color: ${theme.colors.secondaryMedium};

      &::placeholder {
        color: currentColor;
      }
    }
  `,
  alertSuccess: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.successMedium};

      &:focus-within {
        border: 0.1rem solid ${theme.colors.successMedium};
      }
    }
  `,
  alertWarning: (theme: DefaultTheme) => css`
    ${InputWrapper} {
      border-color: ${theme.colors.warningDark};

      &:focus-within {
        border: 0.1rem solid ${theme.colors.warningDark};
      }
    }
  `
}
