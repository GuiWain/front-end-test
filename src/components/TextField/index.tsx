import {
  useState,
  InputHTMLAttributes,
  SetStateAction,
  Dispatch,
  useEffect
} from 'react'

import * as S from './styles'

export type TextFieldProps = {
  onInputChange?: (value: string) => void
  getValue?: Dispatch<SetStateAction<string>>
  type: 'text' | 'email' | 'password' | 'number'
  label?: string
  initialValue?: string | number
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  error?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  type,
  icon,
  iconPosition = 'left',
  label,
  name,
  initialValue = '',
  error,
  disabled = false,
  onInputChange,
  getValue,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)
  const [typeModifier, setTypeModifier] = useState(type)
  const [iconModifier, setIconModifier] = useState(icon)
  const [iconPositionModifier, setIconPositionModifier] = useState(iconPosition)
  const [alert, setAlert] = useState<string | undefined>(undefined)

  const validateEmail = (value: string) => {
    const validate =
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

    if (value === undefined || value === '') {
      setAlert(undefined)
      setIconModifier(undefined)
      setIconPositionModifier('left')
      return
    }

    if (validate.test(value)) {
      setAlert('Valid email')
      setIconModifier(<S.CheckCircleIcon />)
      setIconPositionModifier('right')
    } else {
      setAlert('Invalid email')
      setIconModifier(undefined)
      setIconPositionModifier('left')
    }
  }

  const showPassword = () => {
    setTypeModifier('text')
    setIconModifier(<S.EyeIcon onClick={() => hiddePassword()} />)
  }

  const hiddePassword = () => {
    setTypeModifier('password')
    setIconModifier(<S.EyeSlashIcon onClick={() => showPassword()} />)
  }

  useEffect(() => {
    if (typeModifier === 'password') {
      setIconModifier(<S.EyeSlashIcon onClick={() => showPassword()} />)
      setIconPositionModifier('right')
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const onChange = (value: string) => {
    if (typeModifier === 'email') {
      validateEmail(value)
    }

    setValue(value)

    !!getValue && getValue(value)
    !!onInputChange && onInputChange(value)
  }

  return (
    <S.Wrapper
      disabled={disabled}
      error={!!error}
      alert={
        !!alert && alert === 'Valid email'
          ? 'alertSuccess'
          : undefined || (!!alert && alert === 'Invalid email')
          ? 'alertWarning'
          : undefined
      }
    >
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!iconModifier && (
          <S.Icon iconPosition={iconPositionModifier}>{iconModifier}</S.Icon>
        )}
        <S.Input
          type={typeModifier}
          onChange={(e) => onChange(e.currentTarget.value)}
          value={value}
          iconPosition={iconPositionModifier}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!error && <S.Error>{error}</S.Error>}
      {!!alert && alert === 'Valid email' && !error ? (
        <S.alertSuccess>{alert}</S.alertSuccess>
      ) : null}
      {!!alert && alert === 'Invalid email' && !error ? (
        <S.AlertWarning>{alert}</S.AlertWarning>
      ) : null}
    </S.Wrapper>
  )
}

export default TextField
