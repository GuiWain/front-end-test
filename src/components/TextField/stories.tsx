import { Story, Meta } from '@storybook/react/types-6-0'
import { Address } from '@styled-icons/entypo/'

import TextField, { TextFieldProps } from '.'

export default {
  title: 'Inputs/TextField',
  component: TextField,
  argTypes: {
    onInput: { action: 'changed' },
    type: {
      control: {
        type: 'radio',
        options: ['text', 'email', 'password', 'number']
      }
    },
    placeholder: {
      type: 'string'
    },
    label: {
      type: 'string'
    },
    name: {
      type: 'string'
    },
    initialValue: {
      type: 'string'
    },
    icon: {
      type: 'JSX.element'
    },
    disabled: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    iconPosition: {
      control: {
        type: 'radio',
        options: ['left', 'right']
      }
    }
  }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

Default.args = {
  type: 'text',
  label: 'Enter your name',
  name: 'name',
  placeholder: 'Placeholder',
  disabled: false
}

export const EmailValidate: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

EmailValidate.args = {
  type: 'email',
  label: 'Your email',
  name: 'email',
  placeholder: 'email@email.com.br',
  disabled: false
}

export const WithIcon: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

WithIcon.args = {
  type: 'text',
  label: 'Your adress',
  name: 'address',
  placeholder: 'Calgary',
  icon: <Address />,
  iconPosition: 'left',
  disabled: false
}

export const Password: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

Password.args = {
  type: 'password',
  label: 'Your password',
  placeholder: 'Your password',
  name: 'password',
  disabled: false
}

export const withError: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)

withError.args = {
  type: 'text',
  label: 'Your adress',
  name: 'address',
  placeholder: 'Toronto',
  icon: <Address />,
  iconPosition: 'left',
  disabled: false,
  error: 'Required field'
}
