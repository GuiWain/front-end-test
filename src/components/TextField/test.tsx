import { render, screen, waitFor } from 'utils/test-utils'
import { Email } from '@styled-icons/material-outlined'
import userEvent from '@testing-library/user-event'

import TextField from '.'

describe('<TextField />', () => {
  it('Renders with Label', () => {
    render(<TextField type="email" label="Label" name="Label" />)

    expect(screen.getByLabelText('Label')).toBeInTheDocument()
  })

  it('Renders without Label', () => {
    render(<TextField type="text" />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('Renders with placeholder', () => {
    render(<TextField type="email" placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('Renders with Icon', () => {
    render(<TextField type="email" icon={<Email data-testid="icon" />} />)

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('Renders with Icon on the right side', () => {
    render(
      <TextField
        type="email"
        icon={<Email data-testid="icon" />}
        iconPosition="right"
      />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 })
  })

  it('Changes its value when typing', async () => {
    const onInputChange = jest.fn()
    render(
      <TextField
        onInputChange={onInputChange}
        label="TextField"
        name="TextField"
        type="text"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onInputChange).toHaveBeenCalledWith(text)
  })

  it('Render with props getValue', async () => {
    const getValue = jest.fn()
    render(
      <TextField
        getValue={getValue}
        label="TextField"
        name="TextField"
        type="text"
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(getValue).toHaveBeenCalled()
    })
  })

  it('Does not changes its value when disabled', async () => {
    const onInputChange = jest.fn()
    render(
      <TextField
        type="text"
        onInputChange={onInputChange}
        label="TextField"
        name="TextField"
        disabled
      />
    )

    const input = screen.getByRole('textbox')
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })

  it('Renders with error', () => {
    const { container } = render(
      <TextField type="text" label="TextField" error="Error message" />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('Is accessible by tab', () => {
    render(<TextField type="email" label="TextField" name="TextField" />)

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('Is not accessible by tab when disabled', () => {
    render(
      <TextField type="email" label="TextField" name="TextField" disabled />
    )

    const input = screen.getByLabelText('TextField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })

  it('should be show/hidde password', () => {
    render(<TextField type="password" label="Password" name="TextField" />)

    const inputPassword = screen.getByLabelText(/Password/i)
    const password = 'atus123456'
    userEvent.type(inputPassword, password)

    userEvent.click(screen.getByTestId('eye-slash-icon'))
    expect(inputPassword).toHaveAttribute('type', 'text')

    userEvent.click(screen.getByTestId('eye-icon'))
    expect(inputPassword).toHaveAttribute('type', 'password')

    userEvent.click(screen.getByTestId('eye-slash-icon'))
    expect(inputPassword).toHaveAttribute('type', 'text')
  })

  it('should render valid email', () => {
    render(<TextField type="email" label="TextField" name="TextField" />)

    const input = screen.getByRole('textbox')

    const text = 'email@email.com.br'
    userEvent.type(input, text)

    expect(screen.getByText('Valid email')).toBeInTheDocument()
  })

  it('should render invalid email', () => {
    render(<TextField type="email" label="TextField" name="TextField" />)

    const input = screen.getByRole('textbox')

    const text = 'email.email.com'
    userEvent.type(input, text)

    expect(screen.getByText('Invalid email')).toBeInTheDocument()
  })

  it('should render default wrapper when email is undefined', async () => {
    render(<TextField type="email" label="TextField" name="TextField" />)

    const input = screen.getByLabelText('TextField')
    const text = 'email@email.com'

    userEvent.type(input, text)
    expect(screen.getByText('Valid email')).toBeInTheDocument()

    await waitFor(() => {
      userEvent.clear(input)
      expect(screen.queryByText('Valid email')).not.toBeInTheDocument()
    })
  })
})
