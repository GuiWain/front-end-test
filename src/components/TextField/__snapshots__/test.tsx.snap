// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Renders with error 1`] = `
.c0 .c2 {
  border-color: #ed3020;
}

.c0 .c2:focus-within {
  border: 0.1rem solid #ed3020;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF 0% 0% no-repeat padding-box;
  border: 0.1rem solid #CCC8C8;
  border-radius: 0.4rem;
  padding: 0 0.8rem;
  height: 3.6rem;
}

.c3:focus-within {
  box-shadow: 0rem 0.4rem 0.6rem #00000014;
  border: 0.1rem solid #3241A1;
}

.c4 {
  color: #4F4C4C;
  font: normal normal 400 1.4rem / 2.8rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  padding: 0.8rem 0;
  padding-left: 0.4rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c4:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4 #EAEAEA inset;
  -webkit-filter: none;
  filter: none;
}

.c1 {
  font: normal normal 600 1.6rem / 4rem 'Open Sans','Roboto','Helvetica Neue',sans-serif;
  -webkit-letter-spacing: -0.006rem;
  -moz-letter-spacing: -0.006rem;
  -ms-letter-spacing: -0.006rem;
  letter-spacing: -0.006rem;
  text-align: left;
  color: #4F4C4C;
  cursor: pointer;
}

.c5 {
  color: #ed3020;
  font-size: 1.2rem;
  margin: 0.5rem 0;
}

<div
  class="c0"
>
  <label
    class="c1"
  >
    TextField
  </label>
  <div
    class="c2 c3"
  >
    <input
      class="c4"
      type="text"
      value=""
    />
  </div>
  <p
    class="c5"
  >
    Error message
  </p>
</div>
`;
